src/hvm.rs

```diff
@@ -1633,34 +1653,40 @@ impl Runtime {
   }
 
   // Restores the saved state. This loads the persisted Rollback list and its heaps.
-  fn restore_state(&mut self, uuids: &[u128]) -> std::io::Result<()> {
+  pub fn restore_state(&mut self) -> std::io::Result<()> {
     for i in 0 .. 10 {
       self.heap[i].clear();
     }
-    for i in 0 .. std::cmp::max(uuids.len(), 8) {
-      self.heap[i + 2].load_buffers(uuids[i])?;
-    }
-    let uuids = util::u8s_to_u128s(&std::fs::read(heap_dir_path().join("_uuids_"))?);
-    fn load_heaps(rt: &mut Runtime, uuids: &[u128], index: usize) -> std::io::Result<Arc<Rollback>> {
-      if index == rt.heap.len() {
-        return Ok(Arc::new(Rollback::Nil));
+    self.nuls = vec![2,3,4,5,6,7,8,9];
+    // for i in 0 .. std::cmp::max(uuids.len(), 8) {
+    //   self.heap[i + 2].load_buffers(uuids[i])?;
+    // }
+    let mut uuids = util::u8s_to_u128s(&std::fs::read(heap_dir_path().join("_uuids_"))?);
+    fn load_heaps(rt: &mut Runtime, uuids: &mut Vec<u128>, index: u64, back: Arc<Rollback>) -> std::io::Result<Arc<Rollback>> {
+      if uuids.is_empty() || rt.nuls.is_empty() {
+        return Ok(back);
       } else {
-        rt.heap[index].load_buffers(uuids[index])?;
-        return Ok(Arc::new(Rollback::Cons {
-          keep: 0,
-          head: index as u64, 
-          tail: load_heaps(rt, &uuids, index + 1)?,
-        }));
+        let uuid = uuids.pop().unwrap();
+        let next = rt.nuls.pop().unwrap();
+        rt.heap[index as usize].load_buffers(uuid)?;
+        rt.curr = index;
+        return load_heaps(
+          rt, 
+          uuids, 
+          next, 
+           Arc::new(Rollback::Cons { keep: 0, head: index, tail: back }
+        ));
```

Comment:
This should match on both `pop`s, do the thing on `Some`-`Some`, return Ok on `None`-X, and panic on `Some`-`None`.

If `nuls` is exausted while we still have heaps to load, something wen't wrong, so, `panic!`.